Integrate latest ECL 2.0 syntax changes

cmark opened the pull request:
A preview build of the upcoming ECL 2.0 syntax is available [here](https://github.com/b2ihealthcare/snomed-ecl).

https://snowowl.atlassian.net/browse/SO-4953

cmark committed:
feat(ecl): bump snomed-ecl dependency to 2.0.0

cmark committed:
feat(ecl): support multivalued string evaluation in concrete values

cmark committed:
fix(ecl): compile errors in description term evaluation

cmark committed:
fix(common): add guards to prevent wrapping single collection values...

...again in another collection in HashMapOptions

cmark committed:
test(ecl): print syntax error additional info for easier debugging

cmark committed:
chore: use `Expressions.bool()` instead of deprecated...

...`Expressions.builder()`

cmark committed:
feat(ecl): support `!=` operator in active and module prop filters

cmark committed:
feat(ecl): support `active/moduleId/effectiveTime` property filters...

...with member domain

Also, removed the parameterized nature of property filter tests.

cmark committed:
feat(ecl): support member field property filters

cmark committed:
feat(index): add convenience method to perform single search on query

cmark committed:
chore(constants): add partially equivalent to association concept

cmark committed:
feat(ecl): support history supplements in ECL evaluation

cmark committed:
feat(ecl): MAX is the default history profile in history supplements

cmark committed:
fix(ecl): missing format argument when evaluating history profiles

apeteri commented on commons/com.b2international.commons/src/com/b2international/commons/options/HashMapOptions.java:
```diff
@@ -130,7 +130,7 @@ public String getString(Enum<?> key) {
 	public final <T> Collection<T> getCollection(String key, Class<T> type) {
 		final Object value = get(key);
-		if (type.isInstance(value)) {
+		if (type.isInstance(value) && (value != null && !Collection.class.isAssignableFrom(value.getClass()))) {
```
`isInstance` guarantees non-null input if it evaluates to `true`, so the `value != null` portion is redundant in all three tests:

> This method is the dynamic equivalent of the Java language `instanceof` operator. The method returns `true` if the specified Object argument is non-null and can be cast to the reference type represented by this Class object without raising a `ClassCastException`. It returns `false` otherwise.

apeteri replied:
What was the original reason for this change? You told me but I forgot 😃 Did a caller request eg. `getCollection(key, List.class)`?

apeteri replied:
(The file is also a candidate for a &copy; date bump.)

cmark replied:
> What was the original reason for this change? You told me but I forgot 😃 Did a caller request eg. `getCollection(key, List.class)`?

If you'd like to get back a generic Object List, then the method would create an unnecessary collection wrapping, resulting in `[[ items ]]`, which is incorrect.

apeteri commented on snomed/com.b2international.snowowl.snomed.common/src/com/b2international/snowowl/snomed/common/SnomedConstants.java:
```diff
@@ -250,6 +250,8 @@ private Concepts() { }
 		public static final String REFSET_SIMILAR_TO_ASSOCIATION = "900000000000529008";
 		public static final String REFSET_WAS_A_ASSOCIATION = "900000000000528000";
+		// introduced in 2022-01-31
+		public static final String REFSET_PARTIALLY_EQUIVALENT_TO_ASSOCIATION = "1186924009";
```
The other historical association reference set which was added but we couldn't figure out earlier is `1186921001|POSSIBLY REPLACED BY association reference set|`.

cmark replied:
I'll add that to the list of constants just in case.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore.tests/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclEvaluationRequestHistorySupplementTest.java:
```diff
@@ -0,0 +1,186 @@
+
+	private static final String SIMPLE_REFSET_ID = RandomSnomedIdentiferGenerator.generateConceptId();
+	
+	private static final String INACTIVE_CONCEPT_1 = RandomSnomedIdentiferGenerator.generateConceptId();
+	private static final String INACTIVE_CONCEPT_2 = RandomSnomedIdentiferGenerator.generateConceptId();
+	private static final String INACTIVE_CONCEPT_3 = RandomSnomedIdentiferGenerator.generateConceptId();
+	private static final String INACTIVE_CONCEPT_4 = RandomSnomedIdentiferGenerator.generateConceptId();
+	private static final String INACTIVE_CONCEPT_5 = RandomSnomedIdentiferGenerator.generateConceptId();
```
Please either add some comments, or rename the constants to indicate which ID corresponds to which historical association reference set. It's hard to correlate the expected values in history profile assertions.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore.tests/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclEvaluationRequestHistorySupplementTest.java:
```diff
@@ -0,0 +1,186 @@
+	public void profile_default() throws Exception {
+		Expression actual = eval("<< " + SUBSTANCE + "  {{ + HISTORY }}"); // this equals with MAX profile
+		Expression expected = SnomedConceptDocument.Expressions.ids(Set.of(SUBSTANCE, SUBSTANCE_CHILD_CONCEPT, INACTIVE_CONCEPT_1, INACTIVE_CONCEPT_2, INACTIVE_CONCEPT_3, INACTIVE_CONCEPT_4, INACTIVE_CONCEPT_5));
+		assertEquals(expected, actual);
```
(Optionally, add a message to these assertions telling the reader which historical association reference sets belong to the profile. They can be found in the specification and the `switch` expression that evaluates the query, but would be good to have here as well, for reference.)

cmark replied:
Done.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclEvaluationRequest.java:
```diff
@@ -871,27 +1054,33 @@ static boolean canExtractIds(Expression expression) {
+				return extractIds(expression);
+			}
+			try {
+				return newHashSet(searcher.search(Query.select(String.class)
```
An opportunity to use the newly introduced `Query#search(Searcher)` 😄

apeteri reviewed (approved):
👍 Please see the added comments.

cmark committed:
chore(common): remove unnecessary null checks and bump copyright

cmark committed:
chore(snomed): add POSSIBLY_REPLACED_BY refset id to constants

cmark committed:
chore(test): improve readability of history supplement tests

cmark committed:
feat(ecl): response with unsupported status code when trying to...

...evaluate memberOf filters with unsupported refsetFieldNames (anything
that is not the `referencedComponentId` at the moment)

cmark committed:
chore: use new `Query.searcher` convenience method in ecl eval

cmark committed:
Merge remote-tracking branch 'origin/8.x' into upgrade/ecl-2.0

cmark committed:
Merge remote-tracking branch 'origin/8.x' into upgrade/ecl-2.0

cmark commented:
Closing in favor of #997.
